FSW/src/util/eventUtil.cpp

```diff
@@ -8,96 +8,141 @@
+// sets new duration
+void TimedEvent::setDuration(unsigned long newDuration) { 
     duration = newDuration;
 }
 
+// checks if the event is invoked
 bool TimedEvent::checkInvoked() {
     // will only return true once
     if (!isInvoked && (millis() >= nextInvokeMillis)) {
-        this->invoke();
+        invoke();
         return true;
     }
     return false;
 }
 
-void TimedEvent::start() { // starts the timed event
+// starts the timed event
+void TimedEvent::start() { 
     isInvoked = false;
     nextInvokeMillis = millis() + duration;
 }
 
-/* - - - - - - RecurringEvent - - - - - - *
- * Usage:
- * A RecurringEvent is a TimedEvent that automatically restarts when invoked.
- *  To start the RecurringEvent, call start().
- *  To change the duration of a RecurringEvent call setDuration(newDuration).
- *  To check if the RecurringEvent is invoked call checkInvoked().
- *      If the time has elapsed, checkInvoked() will return true and automatically restart itself.
- */
-RecurringEvent::RecurringEvent(unsigned long newDuration) : TimedEvent(newDuration) { }; // constructor
+
+/* - - - - - - RecurringEvent - - - - - - */
+// constructor
+RecurringEvent::RecurringEvent(unsigned long newDuration) : TimedEvent(newDuration) { }; 
 
 bool RecurringEvent::checkInvoked() {
     // will return true once and automatically restart
     if (!isInvoked && millis() >= nextInvokeMillis) {
-        this->start();
+        start();
         return true;
     }
     return false;
 }
+
+
+/* - - - - - - AsyncEvent - - - - - - */
+ // default constructor
+AsyncEvent::AsyncEvent() : Event() {
+    maxIterations = -1; // will not trigger automatic stop
+    iteration = 0;
+}
+
+// overloaded constructor to set max iterations
+AsyncEvent::AsyncEvent(int maxIter) : Event() { 
+    maxIterations = maxIter;
+    iteration = 0;
+}
+
+// sets new max number of iterations
+void AsyncEvent::setMaxIter(int maxIter) { 
+    maxIterations = maxIter;
+}
+
+// invokes the event
+void AsyncEvent::invoke() {
+    isInvoked = true;
+}
+
+// checks if the event is invoked
+bool AsyncEvent::checkInvoked() { 
+    if (maxIterations > 0 && iteration >= maxIterations) {
+        stop();
+    } else if (isInvoked) {
+        iteration++;
+    }
+    return isInvoked;
+}
+
+// returns the current iteration
+int AsyncEvent::iter() {
+    return iteration; // will return 1 after first call of checkInvoked()
+}
+
+// de-invokes the event and resets current iteration to 0
+void AsyncEvent::stop() {
+    isInvoked = false;
+    iteration = 0;
+}
+
+// de-invokes the event, current iteration will carry over to next call of invoke()
+void AsyncEvent::pause() {
+    isInvoked = false;
+}
+
+// returns whether the current iteration is the first
+bool AsyncEvent::first() {
+    return (iteration == 1);
+}
+
+// returns whether the current iteration is the last
+bool AsyncEvent::over() {
+    return (iteration == maxIterations);
+}
```

Comment:
would it perhaps make sense for this to be `iteration >= maxIterations` just in case the over check is skipped but the event is iterated? There is an argument for keeping it the way it is since that likely would point to a logic error in the code anyway.

Reply:
The only method that iterates `iteration` is `checkInvoked()`, which will automatically reset the event if
`iteration >= maxIterations`, so there is absolutely no way this scenario can occur.